### ISSUE TITLE
Data reader update error

### ISSUE THREAD

**Issue by ndryden** [LLNL/lbann#114]
AlexNet, 128 nodes/256 ranks, with the training dataset having
```
train_or_test_percent: 0.001
validation_percent: 0.01
```
and the testing dataset having
```
train_or_test_percent: 0.01
validation_percent: 1.0
```
which yields
```
Training using 99.0632% of the training data set, which is 1269 samples.
Validating training using 0.936768% of the training data set, which is 12 samples.
```
and
```
--------------------------------------------------------------------------------
[1] Epoch : stats formated [tr/v/te] iter/epoch = [5/1/2]
            global MB = [ 256/  12/ 256] global last MB = [ 245/  12/ 244]
             local MB = [ 256/  12/ 256]  local last MB = [ 245/  12/ 244]
--------------------------------------------------------------------------------
```
completes training and then I get the following error:
```
LBANN: rank 0 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 488 and there are 500 indices
LBANN: rank 9 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 497 and there are 500 indices
LBANN: rank 10 of model 0 caught error message: /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 498 and there are 500 indices
LBANN: rank 4 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 492 and there are 500 indices
LBANN: rank 6 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 494 and there are 500 indices
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
LBANN: rank 8 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 496 and there are 500 indices
LBANN: rank 11 of model 0 caught error message: /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 499 and there are 500 indices
LBANN: rank 5 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 493 and there are 500 indices
LBANN: rank 7 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 495 and there are 500 indices
LBANN: rank 1 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 489 and there are 500 indices
```

**Comment by ndryden** [comment]
(By "completes training" I mean that this appears to be an issue in testing, not training or validation.)

**Comment by ndryden** [comment]
Actually, this seems to occur with any setting on the amount of test data, including 1.0.

**Comment by bvanessen** [comment]
Yeah I noticed at 2:30 am last night that something I did broke a bunch of things.


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Oct 4, 2017, at 7:50 AM, Nikoli Dryden <notifications@github.com<mailto:notifications@github.com>> wrote:


AlexNet, 128 nodes/256 ranks, with the training dataset having

train_or_test_percent: 0.001
validation_percent: 0.01


and the testing dataset having

train_or_test_percent: 0.01
validation_percent: 1.0


which yields

Training using 99.0632% of the training data set, which is 1269 samples.
Validating training using 0.936768% of the training data set, which is 12 samples.


and

--------------------------------------------------------------------------------
[1] Epoch : stats formated [tr/v/te] iter/epoch = [5/1/2]
            global MB = [ 256/  12/ 256] global last MB = [ 245/  12/ 244]
             local MB = [ 256/  12/ 256]  local last MB = [ 245/  12/ 244]
--------------------------------------------------------------------------------


completes training and then I get the following error:

LBANN: rank 0 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 488 and there are 500 indices
LBANN: rank 9 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 497 and there are 500 indices
LBANN: rank 10 of model 0 caught error message: /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 498 and there are 500 indices
LBANN: rank 4 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 492 and there are 500 indices
LBANN: rank 6 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 494 and there are 500 indices
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
LBANN: rank 8 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 496 and there are 500 indices
LBANN: rank 11 of model 0 caught error message: /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 499 and there are 500 indices
LBANN: rank 5 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 493 and there are 500 indices
LBANN: rank 7 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 495 and there are 500 indices
LBANN: rank 1 of model 0 caught error message:  /usr/workspace/wsa/dryden1/lbann/src/data_readers/data_reader.cpp 216 :: generic data reader update error: the epoch is complete, but not all of the data has been used
 -- current pos = 489 and there are 500 indices
